src/generic/complex_matrices.cc

```diff
@@ -538,6 +541,660 @@ namespace oomph
+    }
+
+    // If the matrix is built, check that it's existing distribution is the
+    // same as the in matrix. Since we'll use the same distribution instead
+    // of completely rebuilding it.
+    if (result_matrix.built() &&
+        (*result_matrix.distribution_pt() != *matrix_in.distribution_pt()))
+    {
+      std::ostringstream error_message;
+      error_message << "The result_matrix is built. "
+                    << "But has a different distribution from matrix_in \n"
+                    << "They need to be the same.\n";
+      throw OomphLibError(
+        error_message.str(), OOMPH_CURRENT_FUNCTION, OOMPH_EXCEPTION_LOCATION);
+    }
+#endif
+
+    // To add the elements of two CRDoubleMatrices, we need to know the union of
+    // the sparsity patterns. This is determined by the column indices.
+    // We add the column indices and values (entries) as a key-value pair in
+    // a map (per row). We then read these out into two column indices and
+    // values vector for the result matrix.
+
+    unsigned nrow_local = this->nrow();
+    Vector<int> res_column_indices;
+    Vector<std::complex<double>> res_values;
+    Vector<int> res_row_start;
+    res_row_start.reserve(nrow_local + 1);
+
+    // The row_start and column_indices
+    const int* i_column_indices = this->column_index();
+    const int* i_row_start = this->row_start();
+    const int* in_column_indices = matrix_in.column_index();
+    const int* in_row_start = matrix_in.row_start();
+
+    // Values from this matrix and matrix_in.
+    const std::complex<double>* i_values = this->value();
+    const std::complex<double>* in_values = matrix_in.value();
+
+    // The first entry in row_start is always zero.
+    res_row_start.push_back(0);
+
+    // Loop through the rows of both matrices and insert the column indices and
+    // values as a key-value pair.
+    for (unsigned row_i = 0; row_i < nrow_local; row_i++)
+    {
+      // Create the map for this row.
+      std::map<int, std::complex<double>> res_row_map;
+
+
+      int i_row_end = i_row_start[row_i + 1];
+      // Insert the column and value pair for this matrix.
+      for (int i = i_row_start[row_i]; i < i_row_end; i++)
+      {
+        res_row_map[i_column_indices[i]] = i_values[i];
+      }
+
+      int in_row_end = in_row_start[row_i + 1];
+      // Insert the column and value pair for in matrix.
+      for (int i = in_row_start[row_i]; i < in_row_end; i++)
+      {
+        res_row_map[in_column_indices[i]] += in_values[i];
+      }
+
+      // Fill in the row start
+      res_row_start.push_back(res_row_start.back() + res_row_map.size());
+
+      // Now insert the key into res_column_indices and value into res_values
+      for (std::map<int, std::complex<double>>::iterator it =
+             res_row_map.begin();
+           it != res_row_map.end();
+           ++it)
+      {
+        res_column_indices.push_back(it->first);
+        res_values.push_back(it->second);
+      }
+    }
```

Comment:
With C++11, this can be tidied significantly:
```cpp

// pre-C++11
for (std::map<int, std::complex<double>>::iterator it =
              res_row_map.begin();
            it != res_row_map.end();
            ++it)
{
  res_column_indices.push_back(it->first);
  res_values.push_back(it->second);
}

// C++11; for-range loop with explicit typing
for (const std::pair<int, std::complex<double>> &entry : res_row_map) {
  res_column_indices.push_back(entry.first);
  res_values.push_back(entry.second);
}

// C++11; for-range loop and auto
for (const auto &entry : res_row_map) {
  res_column_indices.push_back(entry.first);
  res_values.push_back(entry.second);
}
```
Although the last one does not explicitly display the type the entries of `res_row_map`, I think it is sufficiently easy to find; the definition is only `res_row_map` 20 lines above (from the start of the loop).

Reply:
@MatthiasHeilManchester, thoughts?

Reply:
Hmmmm; shudders... I've grown up using the first option, obviously. If we're trying to make this the new default that everybody **should** use, i.e.  if use of the pre-C++11 version is actively discouraged, we really ought to discuss this in a bigger group.

Reply:
Good point. In that case, so that Paul can have his freedom back, should we leave it as is?

Reply:
Think so; it's still legal (and very verbose (good!)) code so unless we declare it to be illegal (like `double a=1;` say) let's just leave such constructs.